### PR TITLE
fix a error about aoco table of tsquery type greenplum-db#15556

### PR DESCRIPTION
**Pull request by HelloYJohn**
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by kainwen** [comment]
Thanks for your contribution. Would you please help  write details in commit message (and copy in PR message) so others can learn and review?

**Comment by HelloYJohn** [comment]
I find that aoco table of tsquery type is false and  aoco table of tsvector type is ok, so I compared the two types of select processes and found that geting the value the two types of select processes and found that geting the values are different(I also compared inserting process and found them are different(I also compared inserting process and found them are similar). I think it's just an oversight.

File: src/test/regress/output/uao_ddl/create_ao_tables.source [hunk 1]
```diff
@@ -323,3 +323,18 @@ select count(*) from sto_heap_10;
...
+          c1 TSVECTOR,
+          c2 TSQUERY,
+          c3 INT
+          ) DISTRIBUTED BY (c3);
+INSERT INTO test_ao_tsquery(c1, c2, c3) VALUES('a fat cat sat on a mat and ate a fat rat'::TSVECTOR, 'fat & rat'::TSQUERY, 1);
+ANALYZE test_ao_tsquery;
+SELECT * FROM test_ao_tsquery;
+                         c1                         |      c2       | c3 
+----------------------------------------------------+---------------+----
+ 'a' 'and' 'ate' 'cat' 'fat' 'mat' 'on' 'rat' 'sat' | 'fat' & 'rat' |  1
```
Comment (adam8157) [inline src/test/regress/output/uao_ddl/create_ao_tables.source:337]: I get the same result without this PR, did I miss something?

```
gpadmin=# BEGIN;
CREATE TABLE test_ao_tsquery(
          c1 TSVECTOR,
          c2 TSQUERY,
          c3 INT
          ) DISTRIBUTED BY (c3);
INSERT INTO test_ao_tsquery(c1, c2, c3) VALUES('a fat cat sat on a mat and ate a fat rat'::TSVECTOR, 'fat & rat'::TSQUERY, 1);
ANALYZE test_ao_tsquery;
SELECT * FROM test_ao_tsquery;
BEGIN
CREATE TABLE
INSERT 0 1
ANALYZE
                         c1                         |      c2       | c3
----------------------------------------------------+---------------+----
 'a' 'and' 'ate' 'cat' 'fat' 'mat' 'on' 'rat' 'sat' | 'fat' & 'rat' |  1
(1 row)

```

Comment (HelloYJohn) [reply src/test/regress/output/uao_ddl/create_ao_tables.source:337]: using aoco table，please

Comment (HelloYJohn) [reply src/test/regress/output/uao_ddl/create_ao_tables.source:337]: create table test_ts3(c2 tsquery, c3 int)WITH (appendonly=true, compresstype=zlib, compresslevel=5,orientation=column) distributed by (c3);
insert into test_ts3(c2 ,c3) select 'fat & rat'::tsquery, 1 ;
select * from test_ts3;

Comment (HelloYJohn) [reply src/test/regress/output/uao_ddl/create_ao_tables.source:337]: SET default_table_access_method=ao_column;
BEGIN; 
CREATE TABLE test_ao_tsquery( c1 TSVECTOR, c2 TSQUERY, c3 INT ) DISTRIBUTED BY (c3); INSERT INTO test_ao_tsquery(c1, c2, c3) VALUES('a fat cat sat on a mat and ate a fat rat'::TSVECTOR, 'fat & rat'::TSQUERY, 1); 
ANALYZE test_ao_tsquery; 
SELECT * FROM test_ao_tsquery;
COMMIT;

File: src/include/tsearch/ts_type.h [hunk 2]
```diff
@@ -232,7 +232,7 @@ typedef TSQueryData *TSQuery;
  * but PG_DETOAST_DATUM_COPY is used for simplicity
  */
 
-#define DatumGetTSQuery(X)			((TSQuery) DatumGetPointer(X))
+#define DatumGetTSQuery(X)			((TSQuery) PG_DETOAST_DATUM(X))
```
Comment (adam8157) [inline src/include/tsearch/ts_type.h:235]: I don't get this line, `tsquery` is a plain type, why do you de-toast it?

File: src/test/regress/input/uao_ddl/create_ao_tables.source [hunk 3]
```diff
@@ -205,3 +205,13 @@ set gp_select_invisible=false;
 select * into sto_heap_10 from sto_uao_8;
 select count(*) from sto_heap_10;
 COMMIT;
+BEGIN;
+CREATE TABLE test_ao_tsquery(
+          c1 TSVECTOR,
+          c2 TSQUERY,
+          c3 INT
+          ) DISTRIBUTED BY (c3);
```
Comment (adam8157) [inline src/test/regress/input/uao_ddl/create_ao_tables.source:213]: Change it to an AOCO table, if the issue only reproduces with that.

**Review by adam8157 — CHANGES_REQUESTED** [review]
Why it's related to the table storage, could you please help me understand? Thanks.

File: src/include/tsearch/ts_type.h (hunk 2 above)
Comment (HelloYJohn) [reply src/include/tsearch/ts_type.h:235]: I gdb sql1 "insert into test_ao_tsquery(c2) values ('fat & rat'::TSQUERY)", sql2"select c2 from test_ao_tsquery", sql3 "insert into test_ao_tsquery(c1) values ('a fat cat sat on a mat and ate a fat rat'::TSVECTOR)" and sql4 "select c1 from test_ao_tsquery". I find inserting and selecting of TSVECTOR and TSQUERY is similar.

aocsam.c:aocs_insert_values(AOCSInsertDesc idesc, Datum *d, bool *null, AOTupleId *aoTupleId)
d is processed as toast type， so I stay consistent with TSVECTOR.


#define DatumGetTSVector(X) ((TSVector) PG_DETOAST_DATUM(X))
so I think TSQuery is similar.

File: src/test/regress/input/uao_ddl/create_ao_tables.source (hunk 3 above)
Comment (HelloYJohn) [reply src/test/regress/input/uao_ddl/create_ao_tables.source:213]: SET default_table_access_method=@amname@; has been set at the begin of file create_ao_tables.source, so test_ao_tsquery is ao/aoco table

File: src/include/tsearch/ts_type.h (hunk 2 above)
Comment (adam8157) [reply src/include/tsearch/ts_type.h:235]: It doesn't look strong at first glance, please allow me some time to follow.

**Comment by HelloYJohn** [comment]
image

compile_and_test_gpdb_without_assert/status failed but compile_and_test_gpdb/status success, it becauses analyze_progress is not unstable? 

Can you help me to retry it ? @adam8157

**Comment by adam8157** [comment]
`varattrib_untoast_ptr_len()` looks like a function to de-toast, but it just analyzes and sets `*tofree` to null if the datum is a “varlena” type.

**Comment by adam8157** [comment]
FYI
```
gpadmin=# select typname from pg_type where typlen = -1 and typstorage = 'p';
  typname
------------
 int2vector
 oidvector
 gtsvector
 tsquery
(4 rows)
```

**Comment by adam8157** [comment]
I believe I got a bit better solution, please help reviewing https://github.com/greenplum-db/gpdb/pull/15575

**Pull request closed without merging**